### PR TITLE
Constant time effect sequence

### PR DESCRIPTION
**Pull request by wokalski**
This is a small update that might potentially improve performance. A run of our (admittedly mediocre) benchmark indicates that there might be a slight improvement in the update performance:

```
┌──────────────────────────────────────────────┬────────────────┬──────────┬─────────────┬─────────────┬────────────┐
│ EffectSequence                               │       Time/Run │  mWd/Run │    mjWd/Run │    Prom/Run │ Percentage │
├──────────────────────────────────────────────┼────────────────┼──────────┼─────────────┼─────────────┼────────────┤
│ First render                                 │ 2_728_723.56ns │ 348.10kw │ 127_640.92w │ 127_640.92w │     86.81% │
│ Execute mount effects                        │ 3_143_436.26ns │  15.02kw │     558.28w │     558.28w │    100.00% │
│ Execute host view updates                    │         3.72ns │          │             │             │            │
│ noop flushPendingUpdates                     │ 1_831_831.60ns │ 314.99kw │  36_808.73w │  36_808.73w │     58.27% │
│ State update without subtree change/Depth: 0 │ 1_676_238.01ns │ 314.99kw │  37_235.92w │  37_235.92w │     53.33% │
│ State update without subtree change/Depth: 1 │ 1_600_215.39ns │ 314.89kw │  38_418.40w │  38_418.40w │     50.91% │
│ State update without subtree change/Depth: 2 │ 1_696_933.47ns │ 314.96kw │  37_615.69w │  37_615.69w │     53.98% │
│ State update without subtree change/Depth: 3 │ 1_582_997.96ns │ 314.98kw │  37_301.09w │  37_301.09w │     50.36% │
│ State update without subtree change/Depth: 4 │ 1_638_649.16ns │ 314.98kw │  37_087.71w │  37_087.71w │     52.13% │
│ State update without subtree change/Depth: 5 │ 1_807_074.80ns │ 314.99kw │  37_915.04w │  37_915.04w │     57.49% │
│ State update with subtree change/Depth: 0    │ 1_616_479.94ns │ 314.99kw │  37_531.57w │  37_531.57w │     51.42% │
│ State update with subtree change/Depth: 1    │ 1_537_343.84ns │ 314.91kw │  37_520.55w │  37_520.55w │     48.91% │
│ State update with subtree change/Depth: 2    │ 2_004_529.19ns │ 314.96kw │  37_282.30w │  37_282.30w │     63.77% │
│ State update with subtree change/Depth: 3    │ 1_857_125.38ns │ 314.98kw │  37_194.82w │  37_194.82w │     59.08% │
│ State update with subtree change/Depth: 4    │ 1_911_270.67ns │ 314.98kw │  37_342.30w │  37_342.30w │     60.80% │
│ State update with subtree change/Depth: 5    │ 1_798_738.51ns │ 314.99kw │  37_223.82w │  37_223.82w │     57.22% │
└──────────────────────────────────────────────┴────────────────┴──────────┴─────────────┴─────────────┴────────────┘
```

```
┌──────────────────────────────────────────────┬────────────────┬──────────┬─────────────┬─────────────┬────────────┐
│ List                                         │       Time/Run │  mWd/Run │    mjWd/Run │    Prom/Run │ Percentage │
├──────────────────────────────────────────────┼────────────────┼──────────┼─────────────┼─────────────┼────────────┤
│ First render                                 │ 2_992_289.24ns │ 391.11kw │ 137_426.20w │ 137_426.20w │     91.83% │
│ Execute mount effects                        │ 3_258_544.54ns │  15.02kw │     560.79w │     560.79w │    100.00% │
│ Execute host view updates                    │         5.02ns │          │             │             │            │
│ noop flushPendingUpdates                     │ 2_141_715.66ns │ 367.89kw │  47_433.86w │  47_433.86w │     65.73% │
│ State update without subtree change/Depth: 0 │ 2_032_747.72ns │ 367.89kw │  47_963.73w │  47_963.73w │     62.38% │
│ State update without subtree change/Depth: 1 │ 2_263_501.23ns │ 367.76kw │  48_054.27w │  48_054.27w │     69.46% │
│ State update without subtree change/Depth: 2 │ 2_985_576.76ns │ 367.85kw │  48_215.98w │  48_215.98w │     91.62% │
│ State update without subtree change/Depth: 3 │ 2_715_824.43ns │ 367.88kw │  47_788.77w │  47_788.77w │     83.34% │
│ State update without subtree change/Depth: 4 │ 2_157_891.94ns │ 367.89kw │  47_748.01w │  47_748.01w │     66.22% │
│ State update without subtree change/Depth: 5 │ 2_421_188.51ns │ 367.89kw │  47_620.97w │  47_620.97w │     74.30% │
│ State update with subtree change/Depth: 0    │ 2_281_970.70ns │ 367.89kw │  47_353.22w │  47_353.22w │     70.03% │
│ State update with subtree change/Depth: 1    │ 2_227_244.98ns │ 367.79kw │  47_102.94w │  47_102.94w │     68.35% │
│ State update with subtree change/Depth: 2    │ 2_057_085.35ns │ 367.87kw │  47_060.12w │  47_060.12w │     63.13% │
│ State update with subtree change/Depth: 3    │ 1_998_742.36ns │ 367.89kw │  47_055.08w │  47_055.08w │     61.34% │
│ State update with subtree change/Depth: 4    │ 2_204_624.99ns │ 367.89kw │  46_574.22w │  46_574.22w │     67.66% │
│ State update with subtree change/Depth: 5    │ 1_981_355.80ns │ 367.89kw │  48_892.57w │  48_892.57w │     60.80% │
└──────────────────────────────────────────────┴────────────────┴──────────┴─────────────┴─────────────┴────────────┘
```

@bryphe it'd be cool if you benchmarked it with Revery projects.

### CODE REVIEW

**Comment by bryphe** [comment]
Wow, this actually improved our `EditorSurface` benchmark significantly for Onivim 2!

__BEFORE:__ (latest `master`)
```
+---------------------------------------------------------------------------------------------------------------------------------------------------------------+
|  BENCHMARK                                         |  ITERATIONS |  TIME               |  MINOR GC  |  MAJOR GC  |  MINOR ALLOC  |  PROMOTED  |  MAJOR ALLOC  |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Buffer: Add lines to empty buffer                 |  1000       |  1.70238304138      |  333       |  67        |  28359        |  2685      |  100003685    |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Buffer: Insert line in middle of small buffer     |  1000       |  0.000365972518921  |  1         |  0         |  248029       |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Buffer: Insert line in middle of large buffer     |  1000       |  0.860999107361     |  401       |  80        |  46349        |  4585      |  110006585    |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Buffer: Clear large buffer                        |  1000       |  3.88622283936e-05  |  1         |  0         |  23029        |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  BufferViewTokenizer: Tokenize line, no tokens     |  1000       |  4.60147857666e-05  |  1         |  0         |  24033        |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  EditorSurface - Rendering: Minimal state          |  100        |  0.000457048416138  |  2         |  0         |  276245       |  1918      |  1918         |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  EditorSurface - Rendering: 1000 Lines state       |  100        |  0.000534772872925  |  2         |  0         |  276245       |  1918      |  1918         |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  EditorSurface - Rendering: 100000 Lines state     |  100        |  0.000433921813965  |  2         |  0         |  276245       |  1918      |  1918         |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  EditorSurface - Layout: 1000 Lines state          |  100        |  0.00110793113708   |  1         |  0         |  77132        |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Reducer: SetEditorSize                            |  1000000    |  0.0623679161072    |  195       |  0         |  51001197     |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Reducer: SetEditorFont                            |  1000000    |  0.0633709430695    |  195       |  0         |  51001197     |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Tokenizer: Tokenize large line, high split count  |  1000       |  5.52871894836      |  116       |  2         |  30001989     |  1226166   |  1226166      |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Tokenizer: Tokenize large line, no splits         |  1000       |  3.61539387703      |  13        |  0         |  3152100      |  1599      |  1599         |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

__AFTER:__ (This PR - 19042fb)
```
+---------------------------------------------------------------------------------------------------------------------------------------------------------------+
|  BENCHMARK                                         |  ITERATIONS |  TIME               |  MINOR GC  |  MAJOR GC  |  MINOR ALLOC  |  PROMOTED  |  MAJOR ALLOC  |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Buffer: Add lines to empty buffer                 |  1000       |  1.64314603806      |  333       |  67        |  28359        |  2685      |  100003685    |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Buffer: Insert line in middle of small buffer     |  1000       |  0.000373840332031  |  1         |  0         |  248029       |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Buffer: Insert line in middle of large buffer     |  1000       |  0.837207078934     |  401       |  80        |  46349        |  4585      |  110006585    |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Buffer: Clear large buffer                        |  1000       |  4.00543212891e-05  |  1         |  0         |  23029        |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  BufferViewTokenizer: Tokenize line, no tokens     |  1000       |  5.00679016113e-05  |  1         |  0         |  24033        |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  EditorSurface - Rendering: Minimal state          |  100        |  0.000412940979004  |  1         |  0         |  260933       |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  EditorSurface - Rendering: 1000 Lines state       |  100        |  0.000396013259888  |  1         |  0         |  260981       |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  EditorSurface - Rendering: 100000 Lines state     |  100        |  0.000448942184448  |  1         |  0         |  260981       |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  EditorSurface - Layout: 1000 Lines state          |  100        |  0.00094199180603   |  1         |  0         |  77132        |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Reducer: SetEditorSize                            |  1000000    |  0.060222864151     |  195       |  0         |  51001197     |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Reducer: SetEditorFont                            |  1000000    |  0.05952501297      |  195       |  0         |  51001197     |  25        |  25           |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Tokenizer: Tokenize large line, high split count  |  1000       |  5.50812292099      |  116       |  2         |  30001989     |  1226166   |  1226166      |
|----------------------------------------------------+-------------+---------------------+------------+------------+---------------+------------+---------------|
|  Tokenizer: Tokenize large line, no splits         |  1000       |  3.61763501167      |  13        |  0         |  3152100      |  1599      |  1599         |
+---------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

**Comment by bryphe** [comment]
Sorry for the poor formatting, but there is a huge improvement in the 'Promoted' and 'Major Allocs' for our `<EditorSurface />` benchmark - from 1918 promoted / major -> 25 promoted / major. 

Impressive improvement - thanks @wokalski !

File: lib/EffectSequence.re [hunk 1]
```diff
@@ -0,0 +1,15 @@
+type t = unit => unit;
+
+let noop = () => ();
```
Comment (anmonteiro) [inline lib/EffectSequence.re]: I think you wanna use `Sys.opaque_identity` here?

Comment (wokalski) [reply lib/EffectSequence.re]: Thanks for the tip. I just tried replacing `() => ()` with `opaque_identity` but the results were mixed, no real difference (worked ~0.01% better or worse than `() => ()` depending on a benchmark.)

**Pull request merged**